feat: Add code linting and formatting standards proposal

fxstein opened the pull request:
## Overview

Proposes establishing automated code linting and formatting to maintain consistent code quality across multiple developers.

## What's Proposed

**New Document:** `docs/developer/linting-proposal.md` (480 lines)

A comprehensive proposal for implementing coding standards with:
- Tool selection and configuration
- 7-phase implementation plan
- Conservative rollout strategy
- Progress tracking

## Tools Selected

### Python
- ~~**Black**: Code formatter~~ (line-length=120)
- ~~**Flake8**: Style checker~~ (lenient configuration)
- ~~**Pylint**: Deferred for future consideration~~
- **Ruff** much faster - replaces Black and Flke8
- **mypy** Type checking

### Other Formats
- **yamllint**: YAML validation
- **markdownlint**: Markdown formatting
- **codespell**: Spell checking

## Key Decisions

1. **Line length:** 120 characters
2. **Initial scope:** __init__.py and plugin_solax.py only (test on subset)
3. **Flake8 rules:** Lenient initially, tighten gradually
4. **Enforcement:** Optional initially, evaluate after Phase 6
5. **Pylint:** Deferred until Black/Flake8 are stable

## Example Implementation Strategy (can be simplified)

### 7-Phase Rollout

1. **Phase 1:** Setup configuration files and manual scripts
2. **Phase 2:** Test on 2 core files (__init__.py, plugin_solax.py)
3. **Phase 3:** Add GitHub Actions (informational, won't block PRs)
4. **Phase 4:** Expand to remaining plugin files
5. **Phase 5:** Add optional pre-commit hooks
6. **Phase 6:** Apply to documentation files
7. **Phase 7:** Evaluate feedback, decide if making required

**Conservative approach:** Start small, validate, expand gradually.

## What's Included

### Configuration Files (Examples)
- `.editorconfig` - IDE consistency
- `pyproject.toml` - Black configuration
- `.flake8` - Linter rules
- `.yamllint` - YAML rules
- `.markdownlint.json` - Markdown rules
- `.codespellrc` - Spelling dictionary
- `.pre-commit-config.yaml` - Pre-commit hooks
- `.github/workflows/lint.yml` - GitHub Actions

### Manual Scripts
- `scripts/lint.sh` - Run all checks on demand
- `scripts/format.sh` - Auto-fix formatting issues

## Benefits

### Code Quality
- Consistent style across all files
- Catch common errors early
- Improve readability

### Developer Experience
- Clear standards for contributors
- Automated feedback in PRs (when enabled)
- Less manual code review time

### Maintenance
- Easier code reviews
- Fewer style-related discussions
- Professional project appearance

## Expected Impact

### Phase 2 (Initial Scope)
- **Files:** 2 (__init__.py ~2,000 lines, plugin_solax.py ~9,700 lines)
- **Black:** Estimated 50-200 whitespace/formatting changes
- **Flake8:** Estimated 100-500 style issues (lenient rules)
- **Risk:** Low (Black is safe, Flake8 lenient)

### Phase 4 (Full Scope)
- **Files:** All plugin files (~15-20 files)
- **Impact:** Scales with number of plugins
- **Strategy:** One plugin at a time, test each

## Next Steps

1. **Review proposal** - Gather feedback from maintainers
2. **Adjust based on feedback** - Tool selection, rules, rollout pace
3. **Phase 1:** Implement configuration files
4. **Phase 2:** Test on 2-file subset
5. **Iterate:** Adjust and expand based on results

## Request for Feedback

Looking for input on:
- Tool selection (Black + Flake8 sufficient?)
- 120 character line length (too long/short?)
- Rollout strategy (too conservative/aggressive?)
- When to make enforcement required (Phase 7)
- Any other concerns or suggestions

## Draft Status

This PR is marked as **draft** to facilitate discussion. No code changes yet - just the proposal document for review and feedback.

Once approved, we can proceed with Phase 1 implementation in subsequent PRs.

fxstein committed:
Add code linting and formatting proposal

Created comprehensive proposal for establishing coding standards:

Tools selected:
- Python: Black + Flake8 (Pylint deferred)
- YAML: yamllint
- Markdown: markdownlint
- Spelling: codespell

Key decisions:
- Line length: 120 characters
- Initial scope: __init__.py and plugin_solax.py only
- Flake8: Lenient configuration, tighten gradually
- Enforcement: Optional initially, evaluate after Phase 6

Implementation plan:
- 7 phases from setup through evaluation
- Start with subset of code to validate approach
- Track progression via phase completion
- Conservative rollout strategy

Includes:
- Complete configuration files for all tools
- GitHub Actions workflow
- Pre-commit hooks configuration
- Manual lint/format scripts
- Expected issues analysis
- Benefits and rollout strategy

infradom commented:
I am in favor... 
Lets hope the overhead does not become too large for very small error corrections.

infradom commented:
For readability, I personally prefer longer line lengths (some init.py lines are currently over 180 characters long), using a shorter line length may change the look of our code significantly.

fxstein committed:
Update linting proposal based on feedback: 180 char line length, emphasize automation

Incorporated feedback from @infradom (PR #1701):

Line length changes:
- Changed from 120 to 180 characters across all tools
- Preserves current code readability
- Respects existing long lines (some 180+ chars in __init__.py)

Automation emphasis:
- Added 'Automation-First Approach' section
- Emphasized zero manual overhead
- Clarified optional nature of all manual tools
- GitHub Actions informational only (won't block)
- Pre-commit hooks optional
- Focus: Help scale development, not create barriers

Key principles:
- Developers focus on logic, not formatting
- Automated checks provide feedback
- Won't block small contributions
- Minimal overhead for trivial corrections

Addresses @infradom's concerns about line length and overhead.

fxstein committed:
Phase 1: Add linting configuration files and scripts

Added automated linting configuration (180-char line length):
- .editorconfig - IDE consistency
- pyproject.toml - Black configuration
- .flake8 - Lenient Python style checking
- .yamllint - YAML validation
- .markdownlint.json - Markdown formatting
- .codespellrc - Spell checking

Added manual scripts (optional use):
- scripts/lint.sh - Check __init__.py and plugin_solax.py
- scripts/format.sh - Auto-fix formatting issues

Scope: Limited to __init__.py and plugin_solax.py for Phase 2 testing

All checks are informational/optional at this stage.
Zero manual overhead - fully automated via CI/CD.

Completes subtask #157.1 (Phase 1)

fxstein committed:
Fix spelling errors found by codespell

Fixed spelling issues in core files:
- __init__.py: exeption → exception, mutliple → multiple
- plugin_solax.py: tyoe → type, and other corrections

These were automatically detected by codespell and fixed.
Some fixes applied automatically, manual fixes for ambiguous cases.

Part of Phase 2: Testing linting on subset of code.

fxstein committed:
Phase 2: Apply Black formatting and fix linting issues

Applied Black formatting to core files:
- __init__.py: Formatting improvements (whitespace, indentation)
- plugin_solax.py: Formatting improvements

Fixed spelling errors (codespell):
- Multiple typos corrected automatically and manually

Tuned Flake8 configuration (lenient mode):
- Added E401, E731, E722, E741, E266 to ignore list
- Respects existing code patterns (star imports, lambda, bare except)
- All checks now passing

Results:
✅ Black: No issues (2 files formatted)
✅ Flake8: No issues (lenient config)
✅ Codespell: No issues (spelling fixed)

Phase 2 complete: Linting successfully applied to __init__.py and
plugin_solax.py with zero blocking issues.

Completes subtask #157.3 (apply formatting to codebase subset)

fxstein committed:
Phase 3: Add GitHub Actions workflow for automated linting

Created automated linting workflow:
- Black: Code formatting check (informational)
- Flake8: Style check on __init__.py and plugin_solax.py
- yamllint: YAML validation
- markdownlint: Markdown formatting
- codespell: Spell checking

Workflow runs on:
- Every push
- Every pull request

Status: Informational only - won't block PRs
Purpose: Provide automated feedback to contributors

All checks use configurations from Phase 1:
- 180-character line length
- Lenient Flake8 rules
- Respect existing code patterns

Completes Phase 3 of linting rollout.

fxstein committed:
Fix lint workflow: Limit scope to Phase 2 files only

Fixed GitHub Actions lint workflow failures:

Changes:
- Black: Only check __init__.py and plugin_solax.py (Phase 2 scope)
- Flake8: Already scoped correctly
- yamllint: Disabled for Phase 2 (enable in Phase 6)
- markdownlint: Disabled for Phase 2 (enable in Phase 6)
- codespell: Disabled for Phase 2 (enable in Phase 6)

Rationale:
Phase 2 is testing linting on 2 core Python files only.
Documentation linting (YAML, Markdown, spelling) will be
addressed in Phase 6.

This prevents failures from files outside our current scope.

fxstein committed:
Add GitHub issue and PR templates

Created issue templates:
- bug_report.yml: Structured bug reporting with inverter details
- feature_request.yml: Feature suggestions with use cases

Created PR template:
- Type of change checklist
- Testing requirements
- Breaking changes documentation
- Code review checklist

Templates provide:
- Consistent issue/PR format
- Required information collection
- Better triage and review process
- Improved contributor experience

Completes subtasks #157.4 and #157.5

fxstein commented:
These are first examples of bug/feature/PR templates. Just through something together as a start - they could be simplified

fxstein committed:
Add coding standards and contribution workflow documentation

Created coding-standards.md:
- Automated formatting emphasis (Black, Flake8, codespell)
- 180-character line length standard
- Framework vs plugin code separation
- Performance guidelines (< 1 µs per sensor)
- Common code patterns and examples
- What NOT to do

Created contribution-workflow.md:
- Step-by-step contribution guide
- Fork → Branch → Code → Test → Push → PR
- Syncing with upstream
- Common scenarios (new sensor, bug fix, docs, optimization)
- Optional development tools
- Tips for first-time and regular contributors

Both emphasize:
- Zero manual overhead (automation-first)
- Don't worry about formatting (automated)
- Focus on code quality and testing
- Linting helps but doesn't block

Completes subtasks #157.6 and #157.7

fxstein committed:
Add pre-commit hooks configuration (optional for developers)

Created .pre-commit-config.yaml with Phase 2 scope:

Hooks included:
- Black: Code formatting (180-char line length)
- Flake8: Style checking (lenient config)
- codespell: Spell checking
- Pre-commit hooks: Trailing whitespace, EOF, YAML/JSON validation

Scope limited to:
- custom_components/solax_modbus/__init__.py
- custom_components/solax_modbus/plugin_solax.py

Installation (optional for developers):
  pip install pre-commit
  pre-commit install

Usage:
- Runs automatically on git commit
- Can skip with: git commit --no-verify
- Completely optional - GitHub Actions provides same checks

Phase 5 complete: Pre-commit hooks available for developers who
want local automated checks before committing.

Completes subtask #157.2

fxstein commented:
And some pre-commit hooks so the checks run locally and warn or fail (rare) right away before it reaches the PR

fxstein committed:
Add comprehensive developer environment setup guide

Created development-setup.md with HAOS/SSH add-on focus:

Quick Setup:
- Automated via scripts/install-dependencies.zsh
- Installs all dev tools (black, flake8, yamllint, etc.)
- One command setup for contributors

Detailed Workflow:
- Fork and clone instructions
- Development branch workflow
- Hot reload testing (no restart needed)
- Commit and push process
- PR creation

HAOS Specifics:
- SSH add-on configuration
- Alpine Linux considerations (apk, no pip)
- File system layout
- Home Assistant CLI usage

Development Tools:
- Complete tool list with versions
- Manual installation instructions as fallback
- Optional pre-commit hooks setup

Best Practices:
- Small, focused changes
- Test with hot reload
- Let automation handle formatting
- Check GitHub Actions feedback

Links to:
- Coding standards
- Contribution workflow
- Performance testing
- Plugin validation patterns

Makes it easy for new contributors to get started on HAOS!

Part of task #1 (developer tool dependencies documentation)

fxstein committed:
Clarify that todo.ai is optional and personal preference

Updated development-setup.md to label todo.ai as:
- Optional tool (not required for development)
- Personal preference of @fxstein
- Not a project requirement

This ensures contributors understand todo.ai is a workflow choice,
not a mandatory development tool.

fxstein committed:
Remove private ha-api references, clarify todo.ai as optional

Removed references to ha-api (private tool):
- Replaced with standard 'ha' CLI commands
- Use HA UI for integration status
- Standard Home Assistant commands only

Clarified todo.ai as optional:
- Labeled as personal preference of @fxstein
- Not required for development
- Optional workflow tool

Changes ensure documentation uses only publicly available tools
that all contributors can access.

fxstein commented:
Checking in to see if there is any feedback on this. Happy to adjust - just want to avoid this turning stale.

fxstein committed:
Merge main into feature/coding-standards-and-tooling

- Accept upstream changes for __init__.py (computedEntities rename, improved autorepeat logic)
- Accept upstream changes for plugin_solax.py
- Resolves conflicts by taking upstream (main) version

fxstein committed:
Fix Black and Flake8 linting errors

- Apply Black formatting to all Python files (26 files reformatted)
- Fix Flake8 errors:
  - Fix inline comment spacing (E262)
  - Fix missing whitespace around operators (E228, E231)
  - Fix indentation issues (W191, E101)
  - Fix duplicate dictionary key (F601)
  - Add noqa comments for intentional redefinitions (F811)
  - Fix block comment formatting (E265)
  - Remove trailing whitespace (W291)

All PR checks should now pass.

fxstein commented:
Updated the PR Draft with the latest changes. Happy to discuss or close.

wills106 commented:
Sorry I got a bit overwhelmed with the project as a whole, went on an extended holiday and didn't regain much interest after holiday due to reverse culture shock.

wills106 commented:
I need to catch up on the proposed changes. But I do like the detailed PR notes you have submitted on other PR's, makes it easier to understand what's changed and why.

fxstein committed:
Merge branch 'main' into feature/coding-standards-and-tooling

# Conflicts:
#	custom_components/solax_modbus/const.py
#	custom_components/solax_modbus/plugin_solax.py
#	custom_components/solax_modbus/plugin_solax_ev_charger.py

fxstein committed:
Merge branch 'main' into feature/coding-standards-and-tooling

fxstein committed:
Merge branch 'main' into feature/coding-standards-and-tooling

fxstein committed:
chore: fix lint and spelling issues

Normalize line endings in button module and fix typos to satisfy codespell and Black formatting.

fxstein committed:
Merge branch 'main' into feature/coding-standards-and-tooling
